dask/dataframe/io/parquet/fastparquet.py

```diff
@@ -0,0 +1,303 @@
+        """ Read a single piece of a Parquet dataset into a Pandas DataFrame
+
+        This function is called many times in individual tasks
+
+        Parameters
+        ----------
+        fs: FileSystem
+        piece: object
+            This is some token that is returned by Engine.read_metadata.
+            Typically it represents a row group in a Parquet dataset
+        columns: List[str]
+            List of column names to pull out of that row group
+        partitions:
+        categories:
+
+        Returns
+        -------
+        A Pandas DataFrame
+        """
+        pf = kwargs['pf']
+        categories = kwargs['categories']
+        return pf.read_row_group_file(piece, columns, categories)
+
+    @staticmethod
+    def create_metadata(df, fs, path, append=False, partition_on=None,
+                        ignore_divisions=False, **kwargs):
+
+        fs.mkdirs(path)
+        sep = fs.sep
+
+        object_encoding = kwargs.pop("object_encoding", "utf8")
+        if object_encoding == "infer" or (
+                isinstance(object_encoding,
+                           dict) and "infer" in object_encoding.values()
+        ):
+            raise ValueError(
+                '"infer" not allowed as object encoding, '
+                "because this required data in memory."
+            )
+
+        divisions = df.divisions
+        if append:
+            try:
+                # to append to a dataset without _metadata, need to load
+                # _common_metadata or any data file here
+                pf = fastparquet.api.ParquetFile(path, open_with=fs.open,
+                                                 sep=sep)
+            except (IOError, ValueError):
+                # append for create
+                append = False
+        if append:
+            if pf.file_scheme not in ["hive", "empty", "flat"]:
+                raise ValueError(
+                    "Requested file scheme is hive, "
+                    "but existing file scheme is not."
+                )
+            elif (set(pf.columns) != set(df.columns) - set(partition_on)) or (
+                    set(partition_on) != set(pf.cats)
+            ):
+                raise ValueError(
+                    "Appended columns not the same.\n"
+                    "Previous: {} | New: {}".format(pf.columns,
+                                                    list(df.columns))
+                )
+            elif (pd.Series(pf.dtypes).loc[pf.columns] != df[
+                    pf.columns].dtypes).any():
+                raise ValueError(
+                    "Appended dtypes differ.\n{}".format(
+                        set(pf.dtypes.items()) ^ set(df.dtypes.iteritems())
+                    )
+                )
+            else:
+                df = df[pf.columns + partition_on]
+
+            fmd = pf.fmd
+            i_offset = fastparquet.writer.find_max_part(fmd.row_groups)
+
+            if not ignore_divisions:
```

Comment:
@martindurant, do you think we could move this logic to the core, rather than have it in the engines?

Reply:
I guess you could "open" the dataset in core (i.e., load with the same engine but don't compute) and inspect its divisions.